pkg/boot/universalpayload/utilities.go

```diff
@@ -685,18 +677,67 @@ func buildDeviceTreeInfo(buf io.Writer, mem *kexec.Memory, loadAddr uint64, rsdp
 		dt.NewNode("upl-custom", dt.WithProperty(
 			dt.PropertyU64("hoblistptr", loadAddr+tmpHobOffset),
 		)),
+	)), nil
+}
+
+func constructSMBIOS3Node() (*dt.Node, error) {
+	smbiosTableBase, size, err := getSMBIOSBase()
+
+	// According to EDK2 UPL implementation, only SMBIOS3 is supported in FDT.
+	if (err != nil) || (size != getSMBIOS3HdrSize()) {
+		return nil, errors.Join(ErrSMBIOS3NotFound, err)
+	}
+
+	return dt.NewNode("smbios", dt.WithProperty(
+		dt.PropertyString("compatible", "smbios"),
+		dt.PropertyRegion("reg", uint64(smbiosTableBase), uint64(pageSize)),
+	)), nil
+}
+
+func constructReservedMemoryNode(rsdpBase uint64) *dt.Node {
+	var rsvdNodes []*dt.Node
+
+	acpiChildNode := dt.NewNode("acpi", dt.WithProperty(
+		dt.PropertyString("compatible", "acpi"),
+		dt.PropertyRegion("reg", rsdpBase, uint64(pageSize)),
 	))
 
+	rsvdNodes = append(rsvdNodes, acpiChildNode)
+
+	smbios3ChildNode, err := constructSMBIOS3Node()
```

Comment:
would it make sense to use the if a, b := f(); err != nil {} else {}
for here? Just wondering.

Reply:
Make sense. I have updated commit to address this comment. Thanks Ron.